src/Interpreters/join_common.cpp

```diff
@@ -612,11 +622,16 @@ NotJoinedBlocks::NotJoinedBlocks(std::unique_ptr<RightColumnsFiller> filler_,
         setRightIndex(right_pos, result_position);
     }
 
-    if (column_indices_left.size() + column_indices_right.size() + same_result_keys.size() != result_sample_block.columns())
-        throw Exception(ErrorCodes::LOGICAL_ERROR,
-                        "Error in columns mapping in RIGHT|FULL JOIN. Left: {}, right: {}, same: {}, result: {}",
-                        column_indices_left.size(), column_indices_right.size(),
-                        same_result_keys.size(), result_sample_block.columns());
+    /// `result_sample_block` also may contains non unique column names
+    const auto & result_names = result_sample_block.getNames();
+    size_t unique_names_count = std::unordered_set<String>(result_names.begin(), result_names.end()).size();
```

Comment:
It seems it's possible to avoid unnecessary allocations here (getNames()  return value and unordered_set). `Block` internally maintains unordered_map `index_by_name` with column name as a key. So, by exposing `index_by_name` size via an additional method (for example, `uniqueColumnsCount()`) we can achieve the same effect w/o any allocations